File: app/src/main/java/com/wbrawner/simplemarkdown/presentation/MarkdownPresenter.java
package com.wbrawner.simplemarkdown.presentation;

import android.content.Context;
import android.net.Uri;

import com.wbrawner.simplemarkdown.view.MarkdownEditView;
import com.wbrawner.simplemarkdown.view.MarkdownPreviewView;

import java.io.InputStream;
import java.io.OutputStream;

public interface MarkdownPresenter {
    void loadMarkdown(String fileName, InputStream in);
    void loadFromUri(Context context, Uri fileUri);

    void loadMarkdown(String fileName, InputStream in, FileLoadedListener listener,
                      boolean replaceCurrentFile);

    void newFile(String newName);
    void setEditView(MarkdownEditView editView);
    void setPreviewView(MarkdownPreviewView previewView);

    void saveMarkdown(MarkdownSavedListener listener, String name, OutputStream outputStream);
    void onMarkdownEdited();
    void onMarkdownEdited(String markdown);
    String getFileName();
    void setFileName(String name);
    String generateHTML();
    String generateHTML(String markdown);
    String getMarkdown();
    void setMarkdown(String markdown);

    interface FileLoadedListener {
        void onSuccess(String markdown);

        void onError();
    }

    interface MarkdownSavedListener {
        void saveComplete(boolean success);
    }
}

File: app/src/main/java/com/wbrawner/simplemarkdown/presentation/MarkdownPresenter.kt
package com.wbrawner.simplemarkdown.presentation

import android.content.Context
import android.net.Uri

import java.io.InputStream
import java.io.OutputStream

interface MarkdownPresenter {
    var fileName: String
    var markdown: String
    var editView: MarkdownEditView?
    var previewView: MarkdownPreviewView?
    suspend fun loadFromUri(context: Context, fileUri: Uri): String?
    suspend fun loadMarkdown(
            fileName: String,
            `in`: InputStream,
            replaceCurrentFile: Boolean = true
    ): String?
    fun newFile(newName: String)
    suspend fun saveMarkdown(name: String, outputStream: OutputStream): Boolean
    fun onMarkdownEdited(markdown: String? = null)
    fun generateHTML(markdown: String = ""): String
}

interface MarkdownEditView {
    var markdown: String
    fun setTitle(title: String)

    fun onFileSaved(success: Boolean)

    fun onFileLoaded(success: Boolean)
}

interface MarkdownPreviewView {
    fun updatePreview(html: String)
}


